### PR TITLE
 Improved logging for auth. Added Bad Excl List.

### PR DESCRIPTION
**Pull request by mafranks**
Improved logging for authentication surrounding the recent SecureX API additions.
Added a Bad Exclusions List button and section by request.

### CODE REVIEW

File: bad_exclusions_list.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+bad_exclusions_list = [
+    "THIS IS NOT A LIST OF EXCLUSIONS IN YOUR ENVIRONMENT, BUT A LIST OF EXCLUSIONS WE RECOMMEND AGAINST!",
+    "For the most up to date list and additional information refer to",
+    "https://www.cisco.com/c/en/us/support/docs/security/amp-endpoints/213681-best-practices-for-amp-for-endpoint-excl.html",
+    "",
```
Comment (bmacer) [inline bad_exclusions_list.py:5]: empty exclusion shouldn't be possible, no?  also are 2-4 supposed to be comments?

File: bad_exclusions_list.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    "C:/Windows/Prefetch/*",
+    "C:/Windows/System32/Spool",
+    "C:/Windows/System32/CatRoot2",
+    "C:/Windows/Temp",
+    "C:/Windows/Temp/",
+    "C:/Windows/Temp/*",
+    "C:/Program Files/<company name>/",
+    "C:/Program Files (x86)/<company name>/",
+    "C:/Users/<UserProfileName>/AppData/Local/Temp/",
+    "C:/Users/<UserProfileName>/AppData/LocalLow/Temp/"
```
Comment (bmacer) [inline bad_exclusions_list.py:111]: Wouldn't these exclusions show as C:\\Users (double-backslashes, not forward slashes?)

File: data.py [hunk 3]
```diff
@@ -554,8 +560,11 @@ def parse_xml(self, path=r"C:/Program Files/Cisco/AMP/policy.xml"):
         try:
             root = self.get_root(path)
         except OSError as e:
-            root = self.pull_policy_from_sx()
-
+            if self.api_cred_valid == True:
+                root = self.pull_policy_from_sx()
+            else:
+                logging.debug("Unable to pull policy due to invalid Secure Endpoint API credentials.")
+                exit("Unable to pull policy due to invalid Secure Endpoint API credentials.")
```
Comment (bmacer) [inline data.py:567]: I like it better doing all error checks first, then deal with happy state...
```python
if self.api_cred_valid == False:
    logging.debug("Unable to pull policy due to invalid Secure Endpoint API credentials.")
    exit("Unable to pull policy due to invalid Secure Endpoint API credentials.")
root = self.pull_policy_from_sx()
```

Comment (bmacer) [reply data.py:567]: also should this be .debug or .warning or .error instead?

**Review by bmacer — CHANGES_REQUESTED** [review]

File: bad_exclusions_list.py (hunk 1 above)
Comment (mafranks) [reply bad_exclusions_list.py]: Moved these comments to separate elements like they should have been from the start.
image

File: bad_exclusions_list.py (hunk 2 above)
Comment (mafranks) [reply bad_exclusions_list.py]: Changed these to \\

File: data.py (hunk 3 above)
Comment (mafranks) [reply data.py:565]: Changed the exit criteria comments to .info and the default logging to INFO

**Comment by bmacer** [comment]
1. Failures fail silently (missing stuff in .env file).  Would be much better if there was a pop-up saying "failure, please check logs" at least.  If you want this in a separate issue, let me know.
2. "For more info check this URL" link is not highlightable/copyable.  This is obscene.

File: popups.py [hunk 4]
```diff
@@ -519,6 +520,31 @@ def diag_failed_popup():
...
+def bad_exclusions_popup():
+    '''
+    Show information on exclusions that are not recommended by Cisco.
+    '''
+
+    bad_exclusions_list_text = "\n".join([str(x) for x in bad_exclusions_list])
+    layout = [
+        [sg.Text("THIS IS NOT A LIST OF EXCLUSIONS IN YOUR ENVIRONMENT, BUT A LIST OF EXCLUSIONS WE RECOMMEND AGAINST!")],
+        [sg.Text("For the most up to date list and additional information refer to")],
+        [sg.Text("https://www.cisco.com/c/en/us/support/docs/security/amp-endpoints/213681-best-practices-for-amp-for-endpoint-excl.html")],
```
Comment (bmacer) [inline popups.py:532]: This isn't clickable, can't even be highlighted

**Review by bmacer — CHANGES_REQUESTED** [review]
Failures for missing stuff in .env file should have a popup instead of crashing silently.

File: popups.py (hunk 4 above)
Comment (mafranks) [reply popups.py]: Added hyperlink.  Added popups for authentication errors before exiting the program.

**Review by bmacer — APPROVED** [review]
lgtm

**Pull request merged**